src/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -528,42 +523,67 @@ private static string FormatBigIntegerToHexString(BigInteger value, char format,
                 // [07..00] drop the high 0 as the two's complement positive number remains clear
                 bool clearHighF = false;
                 byte head = bits[cur];
+
                 if (head > 0xF7)
                 {
                     head -= 0xF0;
                     clearHighF = true;
                 }
+
                 if (head < 0x08 || clearHighF)
                 {
                     // {0xF8-0xFF} print as {8-F}
                     // {0x00-0x07} print as {0-7}
-                    fmt = string.Format(CultureInfo.InvariantCulture, "{0}1", format);
-                    sb.Append(head.ToString(fmt, info));
+                    sb.Append(head < 10 ?
+                        (char)(head + '0') :
+                        format == 'X' ? (char)((head & 0xF) - 10 + 'A') : (char)((head & 0xF) - 10 + 'a'));
                     cur--;
                 }
             }
+
             if (cur > -1)
             {
-                fmt = string.Format(CultureInfo.InvariantCulture, "{0}2", format);
-                while (cur > -1)
+                if (format == 'x')
                 {
-                    sb.Append(bits[cur--].ToString(fmt, info));
+                    while (cur > -1)
+                    {
+                        byte b = bits[cur--];
+                        sb.Append(GetLowerCaseHexValue(b >> 4));
+                        sb.Append(GetLowerCaseHexValue(b & 0xF));
+                    }
+                }
+                else
+                {
+                    while (cur > -1)
+                    {
+                        byte b = bits[cur--];
+                        sb.Append(GetUpperCaseHexValue(b >> 4));
+                        sb.Append(GetUpperCaseHexValue(b & 0xF));
+                    }
                 }
             }
+
             if (digits > 0 && digits > sb.Length)
             {
-                // Insert leading zeros.  User specified "X5" so we create "0ABCD" instead of "ABCD"
-                sb.Insert(0, (value._sign >= 0 ? ("0") : (format == 'x' ? "f" : "F")), digits - sb.Length);
+                // Insert leading zeros, e.g. user specified "X5" so we create "0ABCD" instead of "ABCD"
+                sb.Insert(
+                    0,
+                    value._sign >= 0 ? '0' : (format == 'x') ? 'f' : 'F',
+                    digits - sb.Length);
             }
-            return sb.ToString();
+
+            return sb.GetString();
         }
 
+        private static char GetUpperCaseHexValue(int i) => i < 10 ? (char)(i + '0') : (char)(i - 10 + 'A');
```

Comment:
might be clearer to just manually inline both of these.

Reply:
Each is used twice. Why write the code twice?